📖clusterctl: more docs (#2266)
kubernetes-sigs/cluster-api: docs/book/src/clusterctl/commands/config-cluster.md

# clusterctl config cluster

> The `clusterctl create config` command returns a YAML template for creating a workload cluster.

Reviewer: create config?

Author: ooooop... lapsus

selects a cluster template from the `aws` provider's repository. 

In case there is more than one infrastructure provider, the following syntax can be used to select which infrastructure
provider to use for the workload cluster:

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --infrastructure:aws > my-cluster.yaml
```

or

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --infrastructure:aws:v0.4.1 > my-cluster.yaml
```

### Flavors

The infrastructure provider authors can provide different type of cluster templates, or flavors; use the `--flavor` flag
to specify which flavor to use; e.g.

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --flavor high-availabilty > my-cluster.yaml
```

Please refer to the providers documentation for more info about available flavors.

### Alternative source for cluster templates

clusterctl uses the provider's repository as a primary source for cluster templates; the following alternative sources 
for cluster templates can be used as well:

#### ConfigMaps

> Use the `--from-config-map` flag to read cluster templates stored in a Kubernetes ConfigMap; e.g.

Reviewer 2: `--from-config-map` is currently not a supported flag. Is there another PR adding this functionality in?

Author: 😄 https://github.com/kubernetes-sigs/cluster-api/pull/2265

Reviewer 2: We should document the other flags `--from-config-map-key` and `--from-config-map-namespace`. I was testing this out and was about to open issues for those scenarios but then I saw the other flags in the help output. 😄

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --from-config-map my-templates > my-cluster.yaml
```

#### GitHub or local file system folder

Use the `--from` flag to read cluster templates stored in a GitHub repository or in a local file system folder; e.g.

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
   --from https://github.com/my-org/my-repository/blob/master/my-template.yaml > my-cluster.yaml
```

or

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \

>    --from ~/my-template.yaml > my-cluster.yaml

Reviewer 2: This is currently not a flag. Is there another PR adding this functionality in?

Author: 😄  https://github.com/kubernetes-sigs/cluster-api/pull/2265

# clusterctl config cluster

> The `clusterctl create config` command returns a YAML template for creating a workload cluster.

Reviewer 3: ```suggestion
The `clusterctl create config` command returns a YAML template for creating a workload cluster,
which you can store locally, customize, and then finally apply it to start provisioning the
workload cluster.
```

> Store it locally, eventually customize it, and the apply it to start provisioning the workload cluster.

Reviewer 3: ```suggestion
```

For example

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 > my-cluster.yaml
```

Creates a YAML file named `my-cluster.yaml` with a predefined list of Cluster API objects; Cluster, Machines,
Machine Deployments, etc. to be deployed in the current namespace (in case, use the `--target-namespace` flag to 
specify a different target namespace).

> Then, the file can be eventually modified using your editor of choice; when ready, run the following command

Reviewer 3: ```suggestion
Then, the file can be modified using your editor of choice; when ready, run the following command
```

to apply the cluster manifest.

```
kubectl apply -f my-cluster.yaml
```

### Selecting the infrastructure provider to use

The `clusterctl config cluster` command uses smart defaults in order to simplify the user experience; in the example above,
it detects that there is only an `aws` infrastructure provider in the current management cluster and so it automatically 
selects a cluster template from the `aws` provider's repository. 

> In case there is more than one infrastructure provider, following syntax can be used to select which infrastructure

Reviewer 3: ```suggestion
In case there is more than one infrastructure provider, the following syntax can be used to select which infrastructure
```

provider to use for the workload cluster:

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --infrastructure:aws > my-cluster.yaml
```

or

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --infrastructure:aws:v0.4.1 > my-cluster.yaml
```

### Flavors

The infrastructure provider authors can provide different type of cluster templates, or flavors; use the `--flavor` flag

> to specify the which flavor to use; e.g.

Reviewer 3: ```suggestion
to specify which flavor to use; e.g.
```

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --flavor high-availabilty > my-cluster.yaml
```

Please refer to the providers documentation for more info about available flavors.

### Alternative source for cluster templates

> clusterctl uses the provider's repository as a primary source for cluster templates; following alternative sources 

Reviewer 3: ```suggestion
clusterctl uses the provider's repository as a primary source for cluster templates; the following alternative sources 
```

for cluster templates can be used as well:

#### ConfigMaps

> Use the `--from-config-map` flag to read cluster templates stored in Kubernetes ConfigMap; e.g.

Reviewer 3: ```suggestion
Use the `--from-config-map` flag to read cluster templates stored in a Kubernetes ConfigMap; e.g.
```

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
    --from-config-map my-templates > my-cluster.yaml
```

#### GitHub or local file system folder

> Use the `--from` flag to read cluster templates stored into a GitHub repository or in a local file system folder; e.g.

Reviewer 3: ```suggestion
Use the `--from` flag to read cluster templates stored in a GitHub repository or in a local file system folder; e.g.
```

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
   --from https://github.com/my-org/my-repository/blob/master/my-template.yaml > my-cluster.yaml
```

or

```
clusterctl config cluster my-cluster --kubernetes-version v1.16.3 \
   --from ~/my-template.yaml > my-cluster.yaml
```

### Variables

> If the select cluster template expects some environment variables, user should ensure those variables are set in advance.

Reviewer 3: ```suggestion
If the selected cluster template expects some environment variables, you should ensure those variables are set in advance.
```

> e.g. if the `AWS_CREDENTIALS` variable is expected for a cluster template targeting the `aws` infrastructure, user

Reviewer 3: ```suggestion
e.g. if the `AWS_CREDENTIALS` variable is expected for a cluster template targeting the `aws` infrastructure, you
```

should ensure the corresponding environment variable to be set before executing `clusterctl config cluster`.

Please refer to the providers documentation for more info about the required variables or use the 
`clusterctl config cluster --list-variables` flag to get a list of variables names required by a cluster template.

> In case, the [clusterctl configuration](configuration.md) file can be used as alternative to environment variables.

Reviewer 3: ```suggestion
The [clusterctl configuration](configuration.md) file can be used as alternative to environment variables.
```
